Sources/MapboxDirections/Directions.swift

```diff
@@ -158,48 +158,55 @@ open class Directions: NSObject {
             guard let response = possibleResponse, ["application/json", "text/html"].contains(response.mimeType) else {
-                completionHandler(nil, nil, .invalidResponse)
+                let response = RouteResponse(code: nil, message: nil, error: .invalidResponse(possibleResponse), uuid: nil, routes: nil, waypoints: nil)
+                completionHandler(response)
                 return
             }
             
             guard let data = possibleData else {
-                completionHandler(nil, nil, .noData)
+                let response = RouteResponse(code: nil, message: nil, error: .noData, uuid: nil, routes: nil, waypoints: nil)
+                completionHandler(response)
                 return
             }
             
             if let error = possibleError {
-                completionHandler(nil, nil, .unknown(response: possibleResponse, underlying: error, code: nil, message: nil))
+                let unknownError = DirectionsError.unknown(response: possibleResponse, underlying: error, code: nil, message: nil)
+                let response = RouteResponse(code: nil, message: nil, error: unknownError , uuid: nil, routes: nil, waypoints: nil)
+                completionHandler(response)
                 return
             }
             
             DispatchQueue.global(qos: .userInitiated).async {
                 do {
                     let decoder = DirectionsDecoder(options: options)
-                    let result = try decoder.decode(RouteResponse.self, from: data)
+                    var result = try decoder.decode(RouteResponse.self, from: data)
                     guard (result.code == nil && result.message == nil) || result.code == "Ok" else {
                         let apiError = Directions.informativeError(code: result.code, message: result.message, response: response, underlyingError: possibleError)
+                        result.error = apiError
                         DispatchQueue.main.async {
-                            completionHandler(nil, nil, apiError)
+                            completionHandler(result)
                         }
                         return
                     }
                     
-                    guard let routes = result.routes else {
+                    guard result.routes != nil else {
+                        result.error = .unableToRoute
                         DispatchQueue.main.async {
-                            completionHandler(result.waypoints, nil, .unableToRoute)
+                            completionHandler(result)
                         }
                         return
                     }
                     
-                    self.postprocess(routes, fetchStartDate: fetchStart, responseEndDate: responseEndDate, uuid: result.uuid)
+                    result.postprocess(accessToken: self.accessToken, apiEndpoint: self.apiEndpoint, fetchStartDate: fetchStart, responseEndDate: responseEndDate)
```

Comment:
RouteResponse should read in most of these properties via decoder options. Even `responseEndDate` can probably be set automatically at the end of `RouteResponse(from:)` when there’s a `responseStartDate`.